op-e2e/sgt/sgt_test.go

```diff
@@ -0,0 +1,389 @@
+	require.Equal(t, common.Big0.Cmp(txValue), 0)
+	testAccount, testAddr := setUpTestAccount(t, ctx, index, sgt, sgtValue, l2Value)
+
+	// make a simple tx with the testAccount: transfer txValue from testAccount to dummyAddr
+	tx, err := sgt.transferNativeToken(t, ctx, testAccount, dummyAddr, txValue)
+	require.NoError(t, err)
+	receipt, err := wait.ForReceiptOK(ctx, sgt.L2Client, tx.Hash())
+	require.NoError(t, err)
+	gasCost := calcGasFee(receipt)
+
+	// post sgt balance check: it should be 0
+	opts := &bind.CallOpts{Context: ctx}
+	postSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, common.Big0.Cmp(postSgtBalance), 0)
+	// balance invirant check
+	invariantBalanceCheck(t, ctx, sgt, testAddr, gasCost, txValue, sgtValue, l2Value, postSgtBalance)
+}
+
+func fullSGTGasPaymentWithoutNativeBalanceSuccess(t *testing.T, ctx context.Context, index int64, sgtValue *big.Int, l2Value *big.Int, txValue *big.Int, sgt *SgtHelper) {
+	// ensure that native balance is 0
+	require.Equal(t, common.Big0.Cmp(l2Value), 0)
+	// ensure that txValue is 0
+	require.Equal(t, common.Big0.Cmp(txValue), 0)
+	testAccount, testAddr := setUpTestAccount(t, ctx, index, sgt, sgtValue, l2Value)
+
+	// make a simple tx with the testAccount: transfer txValue from testAccount to dummyAddr
+	tx, err := sgt.transferNativeToken(t, ctx, testAccount, dummyAddr, txValue)
+	require.NoError(t, err)
+	receipt, err := wait.ForReceiptOK(ctx, sgt.L2Client, tx.Hash())
+	require.NoError(t, err)
+	gasCost := calcGasFee(receipt)
+
+	// post sgt balance check: sgt should be used as gas first
+	opts := &bind.CallOpts{Context: ctx}
+	postSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, new(big.Int).Add(postSgtBalance, gasCost).Cmp(sgtValue), 0)
+	// balance invirant check
+	invariantBalanceCheck(t, ctx, sgt, testAddr, gasCost, txValue, sgtValue, l2Value, postSgtBalance)
+}
+
+func fullSGTGasPaymentWithNativeBalanceSuccess(t *testing.T, ctx context.Context, index int64, sgtValue *big.Int, l2Value *big.Int, txValue *big.Int, sgt *SgtHelper) {
+	// ensure that native balance is positive
+	require.Equal(t, common.Big0.Cmp(l2Value), -1)
+	// ensure that txValue is 0
+	require.Equal(t, common.Big0.Cmp(txValue), 0)
+	testAccount, testAddr := setUpTestAccount(t, ctx, index, sgt, sgtValue, l2Value)
+
+	// make a simple tx with the testAccount: transfer txValue from testAccount to dummyAddr
+	tx, err := sgt.transferNativeToken(t, ctx, testAccount, dummyAddr, txValue)
+	require.NoError(t, err)
+	receipt, err := wait.ForReceiptOK(ctx, sgt.L2Client, tx.Hash())
+	require.NoError(t, err)
+	gasCost := calcGasFee(receipt)
+
+	// post sgt balance check: sgt should be used as gas first
+	opts := &bind.CallOpts{Context: ctx}
+	postSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, new(big.Int).Add(postSgtBalance, gasCost).Cmp(sgtValue), 0)
```

Comment:
Do we also want to verify that the native balance is unused in this context?

Reply:
It's done in the following invariant check: `preSgtBalance + preL2Balance = gasCost + txValue + postSgtBalance + postL2Balance`. Specifically:
Because `preSgtBalance = postSgtBalance + gasCost` is ensured here, by subtracting the left and right hand of this equation from the above invariant check equation we can derive that:
`preL2Balance = txValue + postSgtBalance `. 
Also, we've required `txValue` to be 0 at first, so `preL2Balance = postSgtBalance ` is implicitly ensured.